cdap-api/src/main/java/co/cask/cdap/api/common/RuntimeArguments.java

```diff
@@ -74,37 +76,85 @@ private RuntimeArguments() {
+   *
+   * Arguments prefixed by "<scopetype>.<name>." have higher precedence than those prefixed by "<scopetype>.*.".
+   * Arguments specified at the global level have lowest precedence.
+   *
    * @param scope The type of the scope
    * @param name The name of the scope, e.g. "myTable"
    * @param arguments the runtime arguments of the enclosing scope
    * @return a map that contains only the keys that start with &lt;scope>.&lt;name>., with that prefix removed.
    */
   public static Map<String, String> extractScope(Scope scope, String name, Map<String, String> arguments) {
-    String prefix = scope + "." + name + ".";
-    return extractPrefix(prefix, arguments);
-  }
-
-  /**
-   * Extract all arguments that start with a prefix, and removes that prefix.
-   * @param prefix The prefix to filter and remove
-   * @param arguments the runtime arguments to extract from
-   * @return a map that contains only the keys that start with the prefix, with the prefix removed.
-   */
-  public static Map<String, String> extractPrefix(String prefix, Map<String, String> arguments) {
     if (arguments == null || arguments.isEmpty()) {
       return arguments;
     }
+
+    String prefix = scope + DOT + name + DOT;
+    String wildCardPrefix = scope + DOT + ASTERISK + DOT;
     Map<String, String> result = Maps.newHashMap();
+    Map<String, String> prefixMatchedArgs = Maps.newHashMap();
+    Map<String, String> wildCardPrefixMatchedArgs = Maps.newHashMap();
+
+    // Group the arguments into different categories based on wild card prefix match, named prefix match or no match
     for (Map.Entry<String, String> entry : arguments.entrySet()) {
-      if (entry.getKey().startsWith(prefix)) {
-        result.put(entry.getKey().substring(prefix.length()), entry.getValue());
+      if (entry.getKey().startsWith(wildCardPrefix)) {
+        // Argument is prefixed with "<scope>.*."
+        wildCardPrefixMatchedArgs.put(entry.getKey().substring(wildCardPrefix.length()), entry.getValue());
+      } else if (entry.getKey().startsWith(prefix)) {
+        // Argument is prefixed with "<scope>.<name>."
+        prefixMatchedArgs.put(entry.getKey().substring(prefix.length()), entry.getValue());
+      } else {
+        // Add global or retainable arguments
+        addGlobalArgument(entry.getKey(), entry.getValue(), scope, result);
       }
     }
+
+    result.putAll(wildCardPrefixMatchedArgs);
+    result.putAll(prefixMatchedArgs);
     return result;
   }
 
   /**
+   * Adds a key to the scoped argument map.
+   * Key is added if -
+   * 1. It is not prefixed with any scope e.g. read.timeout=30
+   * 2. It is prefixed with the retainable scope. In this case key is retained in the argument map till the scope the
+   * key becomes current. e.g. dataset.read.timeout=30. The key is prefixed with the DATASET scope which is retainable.
+   * This key is retained in the argument map till the current scope becomes DATASET at which point it will be removed.
```

Comment:
so this means that is the arguments contain spark.*.debug=true, and the current scope is mapreduce, then it will not be retained? From the perspective of workflow that is probably true. But in general, that is not a property of the scope, but a property of the context where this resolution happens. If the current scope were WORKFLOW (which we do not have yet, but we could in the future), then all the program scopes are retainable, right? 

I think instead of making a property of each scope, you could have a method here in this class: boolean isRetainableScope(Scope currentScope, Scope otherScope). Scopes themselves should be unaware of this.